File: Python/3rdOct.py
import datetime

from datetime import date

def to_date(some_date):
    date1 = datetime.datetime.strptime(some_date,"%d/%m/%Y")
    return date1

def check_function(arrive,depart):
    arrive=to_date(arrive)
    depart=to_date(depart)
    diff=depart-arrive
    days=diff.days
    if days==0:
        return False
    return True

def days_stayed(arrive,depart):
    arrive=to_date(arrive)
    depart=to_date(depart)
    diff=depart-arrive
    days=diff.days
    if days>1:
        return (str(days)+" days")
    elif days==0:
        flag=False
        return flag
    else:
        return (str(days)+" day")

def get_year(visit_date):
    date = datetime.datetime.strptime(visit_date,"%d/%m/%Y")
    return(date.strftime("%Y"))

def get_month(visit_date):
    month = datetime.datetime.strptime(visit_date,"%d/%m/%Y")
    return(month.strftime("%b"))

def get_date(visit_date):
    date = datetime.datetime.strptime(visit_date,"%d/%m/%Y")
    date=(date.strftime("%d"))
    if 4 <= int(date) <= 20 or 24 <= int(date) <= 30:
        suffix= "th"
    else:
        suffix = ["st", "nd", "rd"][int(date) % 10 - 1]
    return str(str(date)+str(suffix))


def itinerary(city_list,visit_date,city_count):
    for i in range(0,city_count-1):
        flag=True
        flag=check_function((visit_date[i]),(visit_date[i+1]))
        if flag==False:
            print("Didn't stay in {} went directly to {}".format(city_list[i],city_list[i+1]))
        else:
            print("Stayed for {} then left for {} on {} {}. ".format(days_stayed(visit_date[i],visit_date[i+1]), city_list[i+1], get_date(visit_date[i+1]), get_month(visit_date[i+1])))

city_count=int(input("Enter number of cities travelled: "))
city_count_dup=city_count
if(city_count>=5 and city_count<=50):
    city_list=[]
    visit_date=[]
    while(city_count>0):
        city,date=input("Enter City Name and Date(DD/MM/YYYY)").split(" ")
        date1 = datetime.datetime.strptime(date,"%d/%m/%Y")
        date1=(date1.strftime("%d/%m/%Y"))
        city_list.append(city)
        visit_date.append(date1)
        city_count-=1
    print("Traveller itinerary in {}, he visited {} on {} {}, ".format(get_year(visit_date[0]), city_list[0], get_date(visit_date[0]), get_month(visit_date[0])))
    itinerary(city_list,visit_date,city_count_dup)
else:
    print("Cities should be between 5 and 50")

File: C++/3rdOct.cpp
#include<bits/stdc++.h>
using namespace std;
struct Date
{
    int year;
    int month;
    int day;
};

const int monthDays[12] = {31, 28, 31, 30, 31, 30,
                           31, 31, 30, 31, 30, 31
                          };

const string monthInThreeLetter[12] = {"Jan", "Feb", "Mar", "Apr", "May","Jun","Jul",
                                       "Aug","Sep","Oct","Nov","Dec"
                                      };

const string suffix[] = {"th ","st ","nd ","rd ","th ","th ","th ","th ","th ","th "};

int countLeapYears(Date date)
{
    int years = date.year;
    if (date.month <= 2)
        years--;

    return years / 4 - years / 100 + years / 400;
}

int getDifference(Date previousDate, Date currentDate)
{
    long int previousDayCount = previousDate.year*365 + previousDate.day;
    for (int i=0; i<previousDate.month - 1; i++)
        previousDayCount += monthDays[i];
    previousDayCount += countLeapYears(previousDate);

    long int currentDayCount = currentDate.year*365 + currentDate.day;
    for (int i=0; i<currentDate.month - 1; i++)
        currentDayCount += monthDays[i];
    currentDayCount += countLeapYears(currentDate);

    return (currentDayCount - previousDayCount);
}


int main()
{
    Date time;
    string city,dateVisited;
    vector<pair<string,Date> > v;
    while(cin>>city>>dateVisited)
    {
        int month;
        int day;
        int year;
        if (sscanf(dateVisited.c_str(), "%d/%d/%d", &day, &month, &year) != 3)
        {
            /// handle error
        }
        else
        {
            time.day = day;
            time.month = month;
            time.year = year;
            v.push_back({city,time});
        }
    }

    cout<<"Traveller itinerary in "<<v[0].second.year<<", he visited "<<v[0].first<<" on "<<v[0].second.day<<suffix[v[0].second.day%10]<< monthInThreeLetter[v[0].second.month-1]<<", ";

    for(int i=1; i<v.size(); i++)
    {
        int daysStayed = getDifference(v[i-1].second,v[i].second);
        if(daysStayed > 0)
        {
            if(daysStayed == 1)
                cout<<"Stayed for "<<daysStayed<<" day then left for "<<v[i].first<<" on "<<v[i].second.day<<suffix[v[i].second.day%10]<<monthInThreeLetter[v[i].second.month-1]<<". ";
            else
                cout<<"Stayed for "<<daysStayed<<" days then left for "<<v[i].first<<" on "<<v[i].second.day<<suffix[v[i].second.day%10]<<monthInThreeLetter[v[i].second.month-1]<<". ";
        }
        else
            cout<<"Didn't stay in "<<v[i-1].first<<" went directly to "<<v[i].first<<". ";
    }
    cout<<endl;
}
